prometheus/promhttp/http.go

```diff
@@ -165,20 +188,32 @@ func HandlerForTransactional(reg prometheus.TransactionalGatherer, opts HandlerO
 		} else {
 			contentType = expfmt.Negotiate(req.Header)
 		}
-		header := rsp.Header()
-		header.Set(contentTypeHeader, string(contentType))
+		rsp.Header().Set(contentTypeHeader, string(contentType))
 
-		w := io.Writer(rsp)
-		if !opts.DisableCompression && gzipAccepted(req.Header) {
-			header.Set(contentEncodingHeader, "gzip")
-			gz := gzipPool.Get().(*gzip.Writer)
-			defer gzipPool.Put(gz)
+		w, encodingHeader, closeWriter, err := NegotiateEncodingWriter(req, rsp, opts.DisableCompression, compressions)
 
-			gz.Reset(w)
-			defer gz.Close()
+		if closeWriter != nil {
+			defer func() {
+				err := closeWriter()
+				if err != nil {
+					if opts.ErrorLog != nil {
+						opts.ErrorLog.Println("error closing writer:", err)
+					}
+				}
+			}()
+		}
+		if err != nil {
```

Comment:
Technically no defer is needed on this error, no? So we can move closeWriter to after err != nil and skip closerWriter if nil check?

Reply:
This one isn't the deferred one, this is the one from NegotiateEncodingWriter. I reordered it for better readability.